k8s: Merge init-container.sh with cilium cleanup

iffyio opened the pull request:
This commit provides the option to only remove BPF
state by passing the flag `--bpf-state-only` to the
`cilium cleanup` command. The idea is to be able to
honor both `clean-cilium-state` and `clean-cilium-bpf-state`
options and replace `init-container.sh` script in the k8s YAMLs.

Fixes #6478

I'm not so sure whether this is the best approach to do this so I'm happy to make suggested changes!
I've skipped updating all the YAML files for now until the PR looks good to merge.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7478)
<!-- Reviewable:end -->

tgraf reviewed (commented):
The direction looks good to me. I'd be happy with this to avoid maintaining a separate container image for the init container. @joestringer: what do you think?

aanm commented on examples/kubernetes/1.13/cilium.yaml:
```diff
@@ -292,8 +292,15 @@ spec:
-      - command:
-        - /init-container.sh
+      - command: ['sh', '-c',
+            'if [ "${CLEAN_CILIUM_STATE}" = "true" ]; then
```
The environment variable `CLEAN_CILIUM_STATE` and `CLEAN_CILIUM_BPF_STATE` can be assigned to the `--force` and `--bpf-state-only` with viper. so we can call `cilium cleanup` instead of detecting if the environment variable is set here.

iffyio replied:
Oh cool, I'll look into this!

aanm reviewed (commented):
looks great, my comment can be addressed in a follow up PR

joestringer commented on cilium/cmd/cleanup.go:
```diff
@@ -56,6 +59,7 @@ const (
 
 func init() {
 	rootCmd.AddCommand(cleanupCmd)
+	cleanupCmd.Flags().BoolVar(&bpfStateOnly, "bpf-state-only", false, "Remove only BPF state")
```
As a matter of style, naming an argument like "--foo-only" means that if we need to add other conditional sets of cleanups in future, then it becomes possible specify "--foo-only --bar-only". In that event, it's confusing whether that's a no-op or whether it's "only" both options.

Could we do something like have a required argument instead, either `cilium cleanup bpf` or `cilium cleanup all`? This would also allow the below code to be something like `if cleanupBPF { ... }` with something at the start that just says `if all { cleanupBPF = true }`.

iffyio replied:
Yes, this makes more sense! Do you see any potential downsides to having them as flags instead of arguments, so that e.g `cilium cleanup bpf` becomes `cilium cleanup --bpf`? Since this will enable the approach that @aanm suggests of binding them to env vars instead of having checks in the YAML.

joestringer replied:
That seems reasonable. My only question would be what does `cilium cleanup` do vs. `cilium cleanup --all`? Would they be the same? (I guess this would retain the previous behaviour, for whatever that is worth)

iffyio replied:
Right, it doesn't seem to fit nicely with this particular case. Intuitively, I could see `cilium cleanup` being the same as `--all` but also as a no-op - essentially requiring a flag to be used with the command. I'll update this PR having it preserve the previous behavior and we can see what works and what doesn't.

joestringer commented on cilium/cmd/cleanup.go:
```diff
@@ -82,37 +86,47 @@ func runCleanup() {
-	fmt.Printf("Warning: Destructive operation. You are about to remove:\n"+
-		"- all BPF maps in %s containing '%s' and '%s'\n"+
-		"- mounted bpffs at %s\n"+
-		"- mounted cgroupv2 at %s\n"+
```
what happened to cgroupv2? The new version of this code doesn't seem to specify this, but the function is still invoked.

iffyio replied:
Thanks for catching this, I had missed it. I'll fix in the follow up

iffyio replied:
Fixed!

joestringer reviewed (changes requested):
I like this change, the only reason I put the container in was because at the time I didn't realise there was a `cilium cleanup` option, and I thought it might be extended in future. This looks simpler and cleaner. 

You can also remove the init container dockerfile and corresponding script from the tree in this PR. (I think they're under something like `contrib/packaging/docker/`)

I have a couple of bits of feedback below.

joestringer reviewed (commented):
This looks great! Were you able to manually verify that the `CLEAN_CILIUM_BPF_STATE` configmap option only cleans up the BPF state and not everything? I'm not sure whether we have automated tests for this..

joestringer commented on examples/kubernetes/templates/v1/cilium-containerd-ds.yaml.sed:
```diff
@@ -135,8 +135,7 @@ spec:
       initContainers:
-      - command:
-        - /init-container.sh
+      - command: ['sh', '-c', 'cilium cleanup -f']
```
These also need to change the `image:` tag for the init container, right?

iffyio replied:
Oh god yes, I only updated this for the 1.13 yaml i tested :man_facepalming: Will fix!

iffyio replied:
Fixed! Also changed running `cilium cleanup -f` with all flags turned off to a no-op.

joestringer reviewed (changes requested):
If `cilium cleanup -f` is always run in the init container, then I think we need to make it a no-op if `CLEAN_CILIUM_STATE` and `CLEAN_CILIUM_BPF_STATE` are both false. Otherwise we'll clear out all state every time Cilium starts, which would break upgrade.

aanm commented on cilium/cmd/cleanup.go:
```diff
@@ -56,17 +70,63 @@ const (
+	cleanupCmd.Flags().BoolVarP(&cleanBPF, bpfFlagName, "", false, "Remove BPF state")
+	cleanupCmd.Flags().BoolVarP(&force, forceFlagName, "f", false, "Skip confirmation")
+
+	bindEnv(allFlagName, cleanCiliumEnvVar)
```
please use the helper that we have for to generate the environment variables
```
import "github.com/cilium/cilium/pkg/option"

option.BindEnv(cleanAll)
option.BindEnv(cleanBPF)
```

For backwards compatibility let's keep the old environment variables (`CLEAN_CILIUM_STATE` and `CLEAN_CILIUM_BPF_STATE`) for a couple releases.

iffyio replied:
Addressed! Let me know what you think. I've only updated the 1.13 cilium.yaml for now.

aanm commented on cilium/cmd/cleanup.go:
```diff
@@ -42,7 +43,20 @@ var cleanupCmd = &cobra.Command{
+)
+
+const (
+	allFlagName   = "all"
```
let's make it more clear with "`cilium cleanup --all-state`" and  "`cilium cleanup --bpf-state`" instead of only `--all` and `--bpf`

iffyio replied:
Fixed!

ianvernon reviewed (approved):
Great work!

iffyio committed:
k8s: Merge initContainer cleanup with cilium cleanup

* Adds two `cilium cleanup` flags `--all-state`
  (env `CILIUM_ALL_STATE`) and `--bpf-state`
  (env `CILIUM_BPF_STATE`) to optionally remove
  all cilium state and only BPF related state respectively.
  The idea is to be able to honor both
  `clean-cilium-state` and `clean-cilium-bpf-state` options
  while replacing the init script logic.
* Replaces cleanup logic in `init-container.sh' with call to
  `cilium cleanup`.
* Changes the base docker image for `cilium-init` to use
  the cilium image in place of busybox in order to have access
  to the cilium binary from `init-container.sh`.
* Updates k8s YAMLs to use new cleanup environment variables.

Fixes #6478

Signed-off-by: ifeanyi <ify1992@yahoo.com>

joestringer reviewed (approved):
This looks fine to me.
